### ISSUE TITLE
[Bug] Pip installation failed in virtual environment

### ISSUE THREAD

**Issue by mr-pmillz** [darkoperator/dnsrecon#268]
**Bug** 🐛 
Installing in a fresh virtual environment fails in python 3.11.8

**To Reproduce**

```shell
git clone https://github.com/darkoperator/dnsrecon.git /opt/dnsrecon && cd /opt/dnsrecon

mkdir ~/pyenv
virtualenv -p python3 ~/pyenv/dnsrecon
source ~/pyenv/dnsrecon/bin/activate

python3 -m pip install -U setuptools wheel
python3 -m pip install -r requirements.txt
python3 -m pip install .
```

**Expected behavior**
Prior to this commit on March 17th 07d631eddd9127838a260aa113e3ac0256505482 , the installation commands above worked.
I have a logfile of it working on March 15th. So my theory is something in the March 17th commit broke something as I first noticed it fail on March 19th.

**Screenshots**
![image](https://github.com/darkoperator/dnsrecon/assets/58376398/999db3f6-f6e5-4243-867f-05438fecf82a)

**System Information (System that tool is running on):**
 - OS: Kali Linux 2024.1
 - Version: `Linux kali 6.6.9-amd64 Kali 6.6.9-1kali1 (2024-01-08) x86_64 GNU/Linux`
 - Python Version: `3.11.8`

**Full Stack Trace**
```shell
┌──(dnsrecon)─(root㉿kali)-[/opt/dnsrecon]
└─# python3 -m pip install .
Processing /opt/dnsrecon
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error

  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [22 lines of output]
      Traceback (most recent call last):
        File "/root/pyenv/dnsrecon/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
        File "/root/pyenv/dnsrecon/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/root/pyenv/dnsrecon/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 118, in get_requires_for_build_wheel
          return hook(config_settings)
                 ^^^^^^^^^^^^^^^^^^^^^
        File "/tmp/pip-build-env-n8488h7q/overlay/lib/python3.11/site-packages/setuptools/build_meta.py", line 325, in get_requires_for_build_wheel
          return self._get_build_requires(config_settings, requirements=['wheel'])
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/tmp/pip-build-env-n8488h7q/overlay/lib/python3.11/site-packages/setuptools/build_meta.py", line 295, in _get_build_requires
          self.run_setup()
        File "/tmp/pip-build-env-n8488h7q/overlay/lib/python3.11/site-packages/setuptools/build_meta.py", line 487, in run_setup
          super().run_setup(setup_script=setup_script)
        File "/tmp/pip-build-env-n8488h7q/overlay/lib/python3.11/site-packages/setuptools/build_meta.py", line 311, in run_setup
          exec(code, locals())
        File "<string>", line 3, in <module>
        File "/opt/dnsrecon/dnsrecon/cli.py", line 45, in <module>
          import dns.flags
      ModuleNotFoundError: No module named 'dns'
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.
```

**Comment by L1ghtn1ng** [comment]
Will have a look at this tomorrow

**Comment by L1ghtn1ng** [comment]
I could not reproduce I did this and it worked for me

* virtualenv -p /usr/bin/python3 dnsrecon-test
* source dnsrecon-test/bin/activate
*  pip3 install -r requirements.txt

That is the recommend way to do it, if you want an easier way to install and do venvs etc take a look at at this also https://github.com/astral-sh/uv

**Comment by mr-pmillz** [comment]
No it's not... 😬  👀 

The issue is:
The setup.py and pyproject.toml in this project are not working together properly and have broken package installation with pip, pipx, etc. so when trying to install via a command such as:
```shell
python3 -m pip install .
# or
pipx install .
# or
pip3 install .
# or
uv pip install -e .
```
it fails..

Your proposed solution of:
installing `requirements.txt` then running with `python3 dnsrecon.py` *works* .. 
**but** is not ideal. Ideally, should be able to install the dnsrecon package correctly...

*https://packaging.python.org/en/latest/discussions/setup-py-deprecated/*
*https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html*

Furthermore, just running:

```shell
python3 -m pip install -r requirements.txt 
```

is incorrect without running setup.py as requirements.txt by itself does not satisfy the data_files installation requirement from setup.py: and since running setup.py from the cli is deprecated.. requirements.txt by itself is... incorrect..

```python
    data_files=[
        (
            '/etc/dnsrecon',
            [
                'namelist.txt',
                'snoop.txt',
                'subdomains-top1mil-20000.txt',
                'subdomains-top1mil-5000.txt',
                'subdomains-top1mil.txt',
            ],
        )
    ],

```

so if run from any dir other than the dnsrecon app root, things will also fail.....so it's not ideal.

Make sense?
Hopefully this clarifies the issue. No hard feelings.. Awesome project, but yeah, please run the original commands I provided so that you can reproduce the issue.

## TL:DR
DNSRecon package installation is currently broken.

**Comment by L1ghtn1ng** [comment]
Making DNSrecon pep517 compliant has been something I have been meaning to do since @branchvincent did it for theHarvester which I am very grateful for, so I just need to take what’s been done there and replicate it ☺️

On 6 Apr 2024, at 04:27, Phil Miller ***@***.***> wrote:

﻿

No it's not... 😬 👀

The issue is:
The setup.py and pyproject.toml in this project are not working together properly and have broken package installation with pip, pipx, etc. so when trying to install via a command such as:

python3 -m pip install .
# or
pipx install .
# or
pip3 install .
# or
uv pip install -e .

it fails..

Your proposed solution of:
installing requirements.txt then running with python3 dnsrecon.py works ..
but is not ideal. Ideally, should be able to install the dnsrecon package correctly...

https://packaging.python.org/en/latest/discussions/setup-py-deprecated/
https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html

Furthermore, just running:

python3 -m pip install -r requirements.txt

is incorrect without running setup.py as requirements.txt by itself does not satisfy the data_files installation requirement from setup.py: and since running setup.py from the cli is deprecated.. requirements.txt by itself is... incorrect..

    data_files=[
        (
            '/etc/dnsrecon',
            [
                'namelist.txt',
                'snoop.txt',
                'subdomains-top1mil-20000.txt',
                'subdomains-top1mil-5000.txt',
                'subdomains-top1mil.txt',
            ],
        )
    ],

so if run from any dir other than the dnsrecon app root, things will also fail.....so it's not ideal.

Make sense?
Hopefully this clarifies the issue. No hard feelings.. Awesome project, but yeah, please run the original commands I provided so that you can reproduce the issue.

TL:DR

DNSRecon package installation is currently broken.

**Comment by mr-pmillz** [comment]
Here's the thing..
Prior to commit: https://github.com/darkoperator/dnsrecon/commit/07d631eddd9127838a260aa113e3ac0256505482

It was pep517 compliant... that is, package installation worked...
That is the issue.

Another way to verify:

```shell
python3 -m pip install git+https://github.com/darkoperator/dnsrecon.git#egg=dnsrecon
```
![image](https://github.com/darkoperator/dnsrecon/assets/58376398/116e12e9-e577-4bc7-8444-6ac3a8e9c73f)

The previous commit 713ba042da1591ce5b69ba36d1d254e622ec2050 worked, and the package could be installed correctly:

![image](https://github.com/darkoperator/dnsrecon/assets/58376398/b3e4292c-cb1c-48d3-aee4-60ff3e715ee3)

```shell
python3 -m pip install git+https://github.com/darkoperator/dnsrecon.git@713ba042da1591ce5b69ba36d1d254e622ec2050#egg=dnsrecon
```

Please take a closer look if you can 🍻

**Comment by branchvincent** [comment]
#270 should fix it

**Comment by mr-pmillz** [comment]
Thanks! @branchvincent @L1ghtn1ng 
💯
